fix: use regular jobs for initializing db

olevski opened the pull request:
/deploy #persist extra-values=initDb.image.repository=registry.dev.renku.ch/tasko.olevski/renku-images/init-db,initDb.image.tag=0.30.2-5.f4125d0.git.2051.h6141ff77

I could not deal with the post-install jobs anymore.

So I replaced all of them. This is what changed:
- no more helm hooks for the jobs, by using the helm revision and a random snippet the jobs are not helm hooks and they still run every time you install, upgrade or rollback
- tested with an internal gitlab and external gitlab
- no more scripts in configmaps 🤬 
- logic makes sense (i.e. when a new secret or database is added it does not get initialized after the helm install finishes but together with everything else)

This is the only way I could think of fixing the crc service database initialization.

In the future we can even expand the python scripts to update the postgres passwords if they change.

~TODO: Add the custom certificate injection to the init jobs.~ I double checked and the only init job that needs custom certs is the one that calls keycloak to make the realms. And the certificates are already there - the other init jobs do not need certificates they do not use https or simply call services internal to the cluster over http or do not use http at all. There is no change in this regard compared to how the old post-install jobs are setup.

olevski commented:
The deployment is broken because our ci action refuses to update the init job images. I will do this by hand later. So this ci deployment is using the old images which exclude all changes in this pr.

olevski commented:
> Does this provide new feature wrt the current status? e.g. if we rotate a DB pwd in the values file will it change it also in the existing DB? If not do you think that is something that we could foresee?

@aledegano this does not add any new features. It just copies the functionality we had. If we wanted to, we could add more features like what you suggest. However if we do this then we should make the init scripts even more structured and probably add tests for them. But yes now it is even easier to add logic in the python file to check if the password for the user has changed and update it. Again if the postgres/admin password is changed there is nothing we can do. But for the other passwords we can use the admin/postgres password to update them.

aledegano commented:
> > Does this provide new feature wrt the current status? e.g. if we rotate a DB pwd in the values file will it change it also in the existing DB? If not do you think that is something that we could foresee?
> 
> @aledegano this does not add any new features. It just copies the functionality we had. If we wanted to, we could add more features like what you suggest. However if we do this then we should make the init scripts even more structured and probably add tests for them. But yes now it is possible to add logic in the python file to check if the password for the user has changed and update it. Again if the postgres/admin password is changed there is nothing we can do. But for the other passwords we can use the admin/postgres password to update them.

ok, thanks for the explanation. It might be worth expanding functionalities yes, but we should also first look into the Postgres operator that also provides most of these operations.

Panaetius commented on scripts/init-db/utils.py:
```diff
@@ -0,0 +1,59 @@
+def gitlab_is_online(url: str) -> int:
+    logging.info("Waiting for gitlab to come online...")
+    res = requests.get(url, timeout=10)
+    if not (res.status_code >= 200 and res.status_code < 400):
```
should we be more strict here? I think for an up and running gitlab it should just return a 200. and e.g. 308 would already be a misconfiguration, I get that if I try just `gitlab.dev.renku.ch/help` (missing `https://`)

Panaetius commented on scripts/init-db/utils.py:
```diff
@@ -0,0 +1,59 @@
+
+@retry(stop=(stop_after_attempt(200) | stop_after_delay(600)), wait=wait_fixed(3), reraise=True)
+def get_k8s_secret(namespace: str, name: str, key: str) -> str:
+    logging.info(f"Reading k8s secret {name} in namesapce {namespace} with key {key}...")
```
namesapce

olevski replied:
I removed this section fully because we do not use it at all.

Panaetius commented on scripts/init-db/utils.py:
```diff
@@ -0,0 +1,59 @@
+
+
+@retry(stop=(stop_after_attempt(200) | stop_after_delay(600)), wait=wait_fixed(3), reraise=True)
+def get_k8s_secret(namespace: str, name: str, key: str) -> str:
```
this isn't used?

olevski replied:
It is not. I thought I would need it because at some point I thought I would need to read the secrets from k8s in some cases. But I do not. I will remove it.

olevski commented on scripts/init-db/gitlab_db_init.py:
```diff
@@ -0,0 +1,105 @@
+    # NOTE: Querying for the base url switches the request to port 443 and the service does not
+    # expose port 443.
+    gitlab_is_online(config.gitlab_url + "/help")
+    with gitlab_db_connection.cursor() as curs:
+        logging.info(f"Cleaning up client id {config.gitlab_oauth_client_id} in gitlab oauth applications...")
+        logging.info(
+            f"Setting up client id {config.gitlab_oauth_client_id} and "
+            f"callback uris {gitlab_callback_uris} in gitlab oauth applications..."
+        )
+        curs.execute(
+            gitlab_oauth_cleanup,
+            {
+                "client_id": config.gitlab_oauth_client_id,
+                "redirect_uris": gitlab_callback_uris,
+                "client_secret": config.gitlab_oauth_client_secret,
+            },
+        )
```
I will move this into a separate class. The dbInit class is general and what we use for all databases.

This is something extremely sketchy we do with the built-in gitlab database. We really should not be doing this in the first place. But since we barely use the built-in gitlab anyhow this rarely runs.

olevski replied:
Moved to a separate class.

olevski commented on scripts/init-db/gitlab_db_init.py:
```diff
@@ -0,0 +1,105 @@
+    gitlab_url: str
+    renku_url: str
+    db_port: int = 5432
+    gitlab_oauth_client_id: str = "renku-ui"
```
You cannot do this though because the fields that have default values cannot preceed fields that do not have

olevski replied:
Yeah that does not work. Let me see if I can reshuffle the order in a different way.

```TypeError: non-default argument 'gitlab_oauth_client_secret' follows default argument```

olevski replied:
This order works. Hopefully that is ok.

```
    db_host: str
    db_admin_username: str
    db_admin_password: str = field(repr=False)
    gitlab_db_username: str
    gitlab_db_name: str
    gitlab_db_password: str = field(repr=False)
    gitlab_url: str
    renku_url: str
    gitlab_oauth_client_secret: str = field(repr=False)
    gitlab_oauth_client_id: str = "renku-ui"
    db_port: int = 5432
```

Panaetius replied:
ah I didn't think about the default values.

olevski commented:
@Panaetius I addressed all your comments. And I also tested locally with an internal gitlab to make sure the changes are all good. So this is good to go!
